In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [2]:
data = pd.read_csv(r"C:\Users\Admin\Data python\hotel_bookings.csv")
pd.set_option('display.max_columns', 32)
data.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     1       1       0.0       0   
3                        0                     1       1       0.0       0   
4                        0                     2       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     GBR         Direct               Direct                  0   
3   BB     GBR      Corporate            Corporate                  0   
4   BB     GBR      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                       0                               0                  C   
1                       0                               0                  C   
2                       0                               0                  A   
3                       0                               0                  A   
4                       0                               0                  A   

  assigned_room_type  booking_changes deposit_type  agent  company  \
0                  C                3   No Deposit    NaN      NaN   
1                  C                4   No Deposit    NaN      NaN   
2                  C                0   No Deposit    NaN      NaN   
3                  A                0   No Deposit  304.0      NaN   
4                  A                0   No Deposit  240.0      NaN   

   days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                     0     Transient   0.0                            0   
1                     0     Transient   0.0                            0   
2                     0     Transient  75.0                            0   
3                     0     Transient  75.0                            0   
4                     0     Transient  98.0                            0   

   total_of_special_requests reservation_status reservation_status_date  
0                          0          Check-Out              2015-07-01  
1                          0          Check-Out              2015-07-01  
2                          0          Check-Out              2015-07-02  
3                          0          Check-Out              2015-07-02  
4                          1          Check-Out              2015-07-03

In [3]:
col = ['agent', 'company', 'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number', 'arrival_date_day_of_month', 'reservation_status_date', 'reservation_status', 'hotel']

hotel_book = data.drop(col, axis= 1)

In [4]:
hotel_book.head()

is_canceled  lead_time  stays_in_weekend_nights  stays_in_week_nights  \
0            0        342                        0                     0   
1            0        737                        0                     0   
2            0          7                        0                     1   
3            0         13                        0                     1   
4            0         14                        0                     2   

   adults  children  babies meal country market_segment distribution_channel  \
0       2       0.0       0   BB     PRT         Direct               Direct   
1       2       0.0       0   BB     PRT         Direct               Direct   
2       1       0.0       0   BB     GBR         Direct               Direct   
3       1       0.0       0   BB     GBR      Corporate            Corporate   
4       2       0.0       0   BB     GBR      Online TA                TA/TO   

   is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  \
0                  0                       0                               0   
1                  0                       0                               0   
2                  0                       0                               0   
3                  0                       0                               0   
4                  0                       0                               0   

  reserved_room_type assigned_room_type  booking_changes deposit_type  \
0                  C                  C                3   No Deposit   
1                  C                  C                4   No Deposit   
2                  A                  C                0   No Deposit   
3                  A                  A                0   No Deposit   
4                  A                  A                0   No Deposit   

   days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                     0     Transient   0.0                            0   
1                     0     Transient   0.0                            0   
2                     0     Transient  75.0                            0   
3                     0     Transient  75.0                            0   
4                     0     Transient  98.0                            0   

   total_of_special_requests  
0                          0  
1                          0  
2                          0  
3                          0  
4                          1

In [5]:
hotel_book.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 23 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   is_canceled                     119390 non-null  int64  
 1   lead_time                       119390 non-null  int64  
 2   stays_in_weekend_nights         119390 non-null  int64  
 3   stays_in_week_nights            119390 non-null  int64  
 4   adults                          119390 non-null  int64  
 5   children                        119386 non-null  float64
 6   babies                          119390 non-null  int64  
 7   meal                            119390 non-null  object 
 8   country                         118902 non-null  object 
 9   market_segment                  119390 non-null  object 
 10  distribution_channel            119390 non-null  object 
 11  is_repeated_guest               119390 non-null  int64  
 12  previous_cancell

In [6]:
hotel_book.describe()

is_canceled      lead_time  stays_in_weekend_nights  \
count  119390.000000  119390.000000            119390.000000   
mean        0.370416     104.011416                 0.927599   
std         0.482918     106.863097                 0.998613   
min         0.000000       0.000000                 0.000000   
25%         0.000000      18.000000                 0.000000   
50%         0.000000      69.000000                 1.000000   
75%         1.000000     160.000000                 2.000000   
max         1.000000     737.000000                19.000000   

       stays_in_week_nights         adults       children         babies  \
count         119390.000000  119390.000000  119386.000000  119390.000000   
mean               2.500302       1.856403       0.103890       0.007949   
std                1.908286       0.579261       0.398561       0.097436   
min                0.000000       0.000000       0.000000       0.000000   
25%                1.000000       2.000000       0.000000       0.000000   
50%                2.000000       2.000000       0.000000       0.000000   
75%                3.000000       2.000000       0.000000       0.000000   
max               50.000000      55.000000      10.000000      10.000000   

       is_repeated_guest  previous_cancellations  \
count      119390.000000           119390.000000   
mean            0.031912                0.087118   
std             0.175767                0.844336   
min             0.000000                0.000000   
25%             0.000000                0.000000   
50%             0.000000                0.000000   
75%             0.000000                0.000000   
max             1.000000               26.000000   

       previous_bookings_not_canceled  booking_changes  days_in_waiting_list  \
count                   119390.000000    119390.000000         119390.000000   
mean                         0.137097         0.221124              2.321149   
std                          1.497437         0.652306             17.594721   
min                          0.000000         0.000000              0.000000   
25%                          0.000000         0.000000              0.000000   
50%                          0.000000         0.000000              0.000000   
75%                          0.000000         0.000000              0.000000   
max                         72.000000        21.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
50%        94.575000                     0.000000                   0.000000  
75%       126.000000                     0.000000                   1.000000  
max      5400.000000                     8.000000                   5.000000

## Explain column meaning

**hotel**：City hotel hoặc Resort Hotel

**lead_time**：Số ngày từ khi đặt phòng đến ngày nghỉ trọ.

**stays_in_weekend_nights**：Số ngày đặt phòng và nghỉ lại (Thứ 7, Chủ Nhật)

**stays_in_week_nights**：Số ngày đặt phòng và nghỉ lại (Thứ 2 ~ Thứ 6)

**country**：Quốc gia.

**market_segment**：Phân loại thị trường(TA：Đại lý du lịch、TO：Điều hành du lịch)

**distribution_channel**：kênh cung cấp(TA：Đại lý du lịch、TO：Điều hành du lịch)

**is_repeated_guest**：có phải repeater hay không

**previous_cancellations**：Đã từng hủy đặt phòng trước đây

**previous_bookings_not_canceled**：chưa từng hủy đặt phòng trước đây

**reserved_room_type**：Loại phòng đặt

**assigned_room_type**：Loại phòng sắp xếp

**booking_changes**：Số lần thay đổi đặt phòng

**agent**：ẩn

**company**：ẩn

**days_in_waiting_list**：Số ngày lấy được xác nhận của khách hàng.

**customer_type**：loại khách hàng

**adr**：Tỉ giá trung bình ngày nghỉ trọ

**required_car_parking_spaces**：Yêu cầu bao nhiêu chỗ đỗ xe

**total_of_special_requests**：Số lượng yêu cầu đặc biệt.

*Đây là bộ dữ liệu trích xuất từ hệ thống Quản lý khách sạn (PMS), vì thế tên khách sạn và đại lý đặt phòng sẽ được ẩn 
đi.*

### Explain for meal

**RO**: Room only

**BB**: Bed & Breakfast

**HB**: Half Board (Breakfast and Dinner normally)

**FB**: Full Board (Beakfast, Lunch and Dinner)

**AI**: All Inclusive (all services of full board plus any others specified in each case)

In [7]:
# Categorical data
col = ['meal', 'country',"market_segment", 'distribution_channel', 'reserved_room_type', 'assigned_room_type',"deposit_type","customer_type"]

for i in hotel_book[col]:
    x = hotel_book[i].value_counts()
    print("Column name is:",i,"and it value is:",x)
    print()

Column name is: meal and it value is: BB           92310
HB           14463
SC           10650
Undefined     1169
FB             798
Name: meal, dtype: int64

Column name is: country and it value is: PRT    48590
GBR    12129
FRA    10415
ESP     8568
DEU     7287
       ...  
NCL        1
NAM        1
DJI        1
SLE        1
MLI        1
Name: country, Length: 177, dtype: int64

Column name is: market_segment and it value is: Online TA        56477
Offline TA/TO    24219
Groups           19811
Direct           12606
Corporate         5295
Complementary      743
Aviation           237
Undefined            2
Name: market_segment, dtype: int64

Column name is: distribution_channel and it value is: TA/TO        97870
Direct       14645
Corporate     6677
GDS            193
Undefined        5
Name: distribution_channel, dtype: int64

Column name is: reserved_room_type and it value is: A    85994
D    19201
E     6535
F     2897
G     2094
B     1118
C      932
H      601
P       12
L    

**meal country**:5

**market_segment**: 177

**market_segment**: 8 

**distribution_channel**: 5 

**reserved_room_type**: 9

**assigned_room_type**: 11

**deposit_type**: 3

**customer_type**: 4

In [8]:
hotel_book.isnull().any()

is_canceled                       False
lead_time                         False
stays_in_weekend_nights           False
stays_in_week_nights              False
adults                            False
children                           True
babies                            False
meal                              False
country                            True
market_segment                    False
distribution_channel              False
is_repeated_guest                 False
previous_cancellations            False
previous_bookings_not_canceled    False
reserved_room_type                False
assigned_room_type                False
booking_changes                   False
deposit_type                      False
days_in_waiting_list              False
customer_type                     False
adr                               False
required_car_parking_spaces       False
total_of_special_requests         False
dtype: bool

In [9]:
hotel_book = hotel_book.dropna()

In [10]:
index_drop = hotel_book[(hotel_book['meal'] == 'Undefined') & (hotel_book['market_segment'] == 'Undefined') & (hotel_book['distribution_channel'] == 'Undefined')]

hotel_book.drop(index_drop.index, inplace = True,axis = 0)

In [11]:
from sklearn.model_selection import train_test_split

train_set, test_set = train_test_split(hotel_book, test_size = 0.2, random_state = 42)

In [12]:
hotel_book.head()

is_canceled  lead_time  stays_in_weekend_nights  stays_in_week_nights  \
0            0        342                        0                     0   
1            0        737                        0                     0   
2            0          7                        0                     1   
3            0         13                        0                     1   
4            0         14                        0                     2   

   adults  children  babies meal country market_segment distribution_channel  \
0       2       0.0       0   BB     PRT         Direct               Direct   
1       2       0.0       0   BB     PRT         Direct               Direct   
2       1       0.0       0   BB     GBR         Direct               Direct   
3       1       0.0       0   BB     GBR      Corporate            Corporate   
4       2       0.0       0   BB     GBR      Online TA                TA/TO   

   is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  \
0                  0                       0                               0   
1                  0                       0                               0   
2                  0                       0                               0   
3                  0                       0                               0   
4                  0                       0                               0   

  reserved_room_type assigned_room_type  booking_changes deposit_type  \
0                  C                  C                3   No Deposit   
1                  C                  C                4   No Deposit   
2                  A                  C                0   No Deposit   
3                  A                  A                0   No Deposit   
4                  A                  A                0   No Deposit   

   days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                     0     Transient   0.0                            0   
1                     0     Transient   0.0                            0   
2                     0     Transient  75.0                            0   
3                     0     Transient  75.0                            0   
4                     0     Transient  98.0                            0   

   total_of_special_requests  
0                          0  
1                          0  
2                          0  
3                          0  
4                          1

In [13]:
X_notprepared_train_set = train_set.drop(["is_canceled"], axis = 1)

# Take the booking has been canceled
Y_train_set = (train_set["is_canceled"] == 1).copy()

In [14]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder

num_attri = ["lead_time", 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'booking_changes', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces']
cat_attri = ['meal', 'country', 'market_segment', 'distribution_channel', 'reserved_room_type', 'assigned_room_type', 'deposit_type', 'customer_type']

full_pipeline = ColumnTransformer([
    ("num", StandardScaler(), num_attri),
    ("cat", OneHotEncoder(), cat_attri)
])

X_prepared_train_set = full_pipeline.fit_transform(X_notprepared_train_set)

In [15]:
X_prepared_train_set = X_prepared_train_set.toarray()

In [17]:
from sklearn.decomposition import PCA

pca = PCA(n_components = 0.95)

X_prepared_train_set_reduced = pca.fit_transform(X_prepared_train_set)

In [18]:
X_prepared_train_set_reduced.shape

(95118, 24)

In [20]:
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import cross_val_predict

sgd_clf = SGDClassifier(random_state = 42)

sgd_clf.fit(X_prepared_train_set_reduced, Y_train_set)

SGDClassifier(alpha=0.0001, average=False, class_weight=None,
              early_stopping=False, epsilon=0.1, eta0=0.0, fit_intercept=True,
              l1_ratio=0.15, learning_rate='optimal', loss='hinge',
              max_iter=1000, n_iter_no_change=5, n_jobs=None, penalty='l2',
              power_t=0.5, random_state=42, shuffle=True, tol=0.001,
              validation_fraction=0.1, verbose=0, warm_start=False)

In [29]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import confusion_matrix

def EvalAcc_Confumatrix(model, X, Y, cv = 5):
    result1 = cross_val_score(model, X, Y, scoring = "accuracy", cv = cv)
    result2 = cross_val_predict(model, X, Y, cv = cv)
    print("score:", result1)
    print("confu_matrix:", confusion_matrix(result2, Y))

In [30]:
EvalAcc_Confumatrix(sgd_clf, X_prepared_train_set_reduced, Y_train_set)

score: [0.77712363 0.77554668 0.77559924 0.78709983 0.78121222]
confu_matrix: [[55118 16215]
 [ 4776 19009]]


In [32]:
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score

def EvalROCcurve_Plot(model, X, Y, cv = 5, method = "decision_function", 
                      styleplot = "seaborn", label = None, title = None):
    result3 = cross_val_predict(model, X, Y, cv = cv, method = method)
    result4 = roc_auc_score(Y, result3)
    fpr, tpr, thresholds = roc_curve(Y, result3)
    with plt.style.context(styleplot):
        fig, ax = plt.subplots(figsize = (12, 8))
        ax.plot(fpr, tpr, linewidth = 2, label = label)
        ax.plot([0, 1], [0, 1], "k--")
        ax.set(xlabel = "Recall", ylabel  = "False Positive Rate", title = title)
    return result3
    return result4

In [ ]:
from sklearn.metrics import roc_curve

fpr, tpr, thresholds = roc_curve(hotel_label_train_set,label_scores_pre)